libnd4j/include/helpers/cpu/MmulHelper.cpp

```diff
@@ -21,117 +21,128 @@
+    const Nd4jLong cLen = vC->lengthOf();
 
-    //         if(flagC) {
-    //             PRAGMA_OMP_SIMD
-    //             for(uint col = 0; col < N; ++col) {
-    //                 if(betaZ)
-    //                     c[col * ldc] += a * b[flagB ? col : col * ldb] + betaZ * c[col * ldc];
-    //                 else
-    //                     c[col * ldc] += a * b[flagB ? col : col * ldb];
-    //             }
-    //         }
-    //         else {
-    //             PRAGMA_OMP_SIMD
-    //             for(uint col = 0; col < N; ++col) {
-    //                 if(betaZ)
-    //                     c[col] += a * b[flagB ? col : col * ldb] + betaZ * c[col];
-    //                 else
-    //                     c[col] += a * b[flagB ? col : col * ldb];
-    //             }
-    //         }
-    //     }
-    // }   
+    const int K = vA->sizeAt(aKaxis);
 
-    auto func = PRAGMA_THREADS_FOR_2D { ;
-        for (auto row = start_x; row < stop_x; row += inc_x) {
-            for (auto col = start_y; col < stop_y; col += inc_y) {
-                T3 *c = flagC ? (C + row + col * ldc) : (C + row * ldc + col);
-                T3 val = 0;
+    auto func = PRAGMA_THREADS_FOR {
 
-                PRAGMA_OMP_SIMD
-                for (uint i = 0; i < K; ++i) {
-                    T3 a = flagA ? *(A + row * lda + i) : *(A + row + i * lda);
-                    T3 b = flagB ? *(B + col + i * ldb) : *(B + col * ldb + i);
-                    val += alphaZ * a * b;
-                }
+        std::vector<Nd4jLong> aCoords(2), bCoords(2), cCoords(2);
 
-                if (betaZ)
-                    *c = val + betaZ * *c;
-                else
-                    *c = val;
+        for (auto i = start; i < stop; ++i) {
+
+            // evaluate C coordinates
+            shape::index2coords(i, cShapeInfo, cCoords.data());
+
+            // evaluate A coordinates
+            aCoords[aMaxis] = cCoords[cMaxis];
+            aCoords[aKaxis] = 0;
+
+            // evaluate B coordinates
+            bCoords[bKaxis] = 0;
+            bCoords[bNaxis] = cCoords[cNaxis];
+
+            auto aOffset = shape::getOffset(aShapeInfo, aCoords.data());
+            auto bOffset = shape::getOffset(bShapeInfo, bCoords.data());
+
+            T3 val = A[aOffset] * B[bOffset];                       // first iteration
+
+            for (uint j = 1; j < K; ++j) {                          // rest iterations
+                aOffset += shape::stride(aShapeInfo)[aKaxis];
+                bOffset += shape::stride(bShapeInfo)[bKaxis];
+                val = val + A[aOffset] * B[bOffset];
             }
+
+            auto cOffset = shape::getOffset(cShapeInfo, cCoords.data());
+
+            if(betaZ)
+                C[cOffset] = alphaZ * val + betaZ * C[cOffset];
+            else
+                C[cOffset] = alphaZ * val;
         }
     };
 
-    samediff::Threads::parallel_for(func, 0, M, 1, 0, N, 1);
+    samediff::Threads::parallel_for(func, 0, cLen);
```

Comment:
in this particular case you probably want to use `parallel_tad` instead of `parallel_for`... we probably should test performance difference here. If you execute `parallel_for` with clen < 1024 you'll still have 1 thread running. Which might lead to suboptimal results in case of skewed matrices. Say [16, 1024] x [1024, 16] = [16, 16]...

So ye, let's test this first.